File: examples/python/customer_usage_sync.py
import json
import os
from datetime import datetime, timedelta, timezone
from pathlib import Path
from urllib.parse import quote, urlencode
from urllib.request import Request, urlopen


def load_local_env() -> None:
    env_path = Path(__file__).with_name(".env")
    if not env_path.exists():
        return

    for raw_line in env_path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue

        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))


def build_window(days: int) -> dict[str, str]:
    to = datetime.now(timezone.utc)
    from_ = to - timedelta(days=days)
    return {
        "from": from_.isoformat().replace("+00:00", "Z"),
        "to": to.isoformat().replace("+00:00", "Z"),
    }


def request_json(*, base_url: str, session_token: str, tenant_id: str, path: str) -> dict:
    request = Request(
        f"{base_url}{path}",
        headers={
            "authorization": f"Bearer {session_token}",
            "x-tenant-id": tenant_id,
        },
    )
    with urlopen(request) as response:  # noqa: S310
        return json.loads(response.read().decode("utf-8"))


def to_ledger_snapshot(customer: dict) -> dict:
    return {
        "customer_id": customer["customer_id"],
        "billable_units": customer["total_tokens"],
        "estimated_cost_usd": customer["total_cost_usd"],
        "request_count": customer["request_count"],
        "project_count": customer["project_count"],
        "sync_note": "Map billable_units to your own credit formula before persisting.",
    }


def to_webhook_ledger_event(payload: dict) -> dict:
    return {
        "customer_id": payload["customer_id"],
        "event_type": payload["event_type"],
        "billable_units": payload["overview"]["total_tokens"],
        "estimated_cost_usd": payload["overview"]["total_cost_usd"],
        "credit_bucket": payload.get("credit_bucket"),
        "remaining_credits": payload.get("remaining_credits"),
        "threshold": payload.get("threshold"),
    }


def main() -> None:
    load_local_env()

    base_url = os.getenv("TOKVERA_API_BASE_URL")
    session_token = os.getenv("TOKVERA_SESSION_TOKEN")
    tenant_id = os.getenv("TOKVERA_TENANT_ID")
    project_id = os.getenv("TOKVERA_PROJECT_ID", "")
    preferred_customer_id = os.getenv("TOKVERA_CUSTOMER_ID", "")
    days = max(1, int(os.getenv("TOKVERA_CUSTOMER_USAGE_DAYS", "30")))

    if not base_url:
        raise RuntimeError("TOKVERA_API_BASE_URL is required")
    if not session_token:
        raise RuntimeError("TOKVERA_SESSION_TOKEN is required")
    if not tenant_id:
        raise RuntimeError("TOKVERA_TENANT_ID is required")

    window = build_window(days)
    query = {"from": window["from"], "to": window["to"]}
    if project_id:
        query["project_id"] = project_id

    list_payload = request_json(
        base_url=base_url,
        session_token=session_token,
        tenant_id=tenant_id,
        path=f"/v1/usage/customers?{urlencode(query)}",
    )

    selected_customer = next(
        (item for item in list_payload["items"] if item["customer_id"] == preferred_customer_id),
        list_payload["items"][0] if list_payload["items"] else None,
    )

    if not selected_customer:
        print("No customer-attributed usage found in this window.")
        return

    detail_payload = request_json(
        base_url=base_url,
        session_token=session_token,
        tenant_id=tenant_id,
        path=f"/v1/usage/customers/{quote(selected_customer['customer_id'], safe='')}?{urlencode(query)}",
    )

    webhook_payload = {
        "schema_version": "customer_usage.v1",
        "event_type": "customer.usage.updated",
        "customer_id": selected_customer["customer_id"],
        "credit_bucket": (
            detail_payload["by_credit_bucket"][0]["group"] if detail_payload["by_credit_bucket"] else None
        ),
        "remaining_credits": None,
        "threshold": None,
        "overview": detail_payload["overview"],
    }

    print("customer usage sync example complete")
    print(
        json.dumps(
            {
                "window": window,
                "selected_customer": selected_customer["customer_id"],
                "ledger_snapshot": to_ledger_snapshot(selected_customer),
                "webhook_projection": to_webhook_ledger_event(webhook_payload),
                "next_step": (
                    "Persist ledger_snapshot in your own billing store and validate webhook_projection "
                    "against your webhook handler before enabling automated customer_usage webhooks."
                ),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: examples/node/customer-usage-sync.mjs
import { loadLocalEnv } from "./env.mjs";

function buildWindow(days) {
  const to = new Date();
  const from = new Date(to.getTime() - days * 24 * 60 * 60 * 1000);
  return {
    from: from.toISOString(),
    to: to.toISOString(),
  };
}

async function requestJson({ baseUrl, sessionToken, tenantId, path }) {
  const response = await fetch(`${baseUrl}${path}`, {
    headers: {
      authorization: `Bearer ${sessionToken}`,
      "x-tenant-id": tenantId,
    },
  });

  if (!response.ok) {
    throw new Error(`Request failed (${response.status}) for ${path}`);
  }

  return response.json();
}

function toLedgerSnapshot(customer) {
  return {
    customer_id: customer.customer_id,
    billable_units: customer.total_tokens,
    estimated_cost_usd: customer.total_cost_usd,
    request_count: customer.request_count,
    project_count: customer.project_count,
    sync_note: "Map billable_units to your own credit formula before persisting.",
  };
}

function toWebhookLedgerEvent(payload) {
  return {
    customer_id: payload.customer_id,
    event_type: payload.event_type,
    billable_units: payload.overview.total_tokens,
    estimated_cost_usd: payload.overview.total_cost_usd,
    credit_bucket: payload.credit_bucket,
    remaining_credits: payload.remaining_credits,
    threshold: payload.threshold,
    metadata: payload.metadata || null,
    metadata_observed_at: payload.metadata_observed_at || null,
  };
}

async function main() {
  loadLocalEnv();

  const baseUrl = process.env.TOKVERA_API_BASE_URL;
  const sessionToken = process.env.TOKVERA_SESSION_TOKEN;
  const tenantId = process.env.TOKVERA_TENANT_ID;
  const projectId = process.env.TOKVERA_PROJECT_ID || "";
  const preferredCustomerId = process.env.TOKVERA_CUSTOMER_ID || "";
  const days = Math.max(1, Number(process.env.TOKVERA_CUSTOMER_USAGE_DAYS || "30") || 30);

  if (!baseUrl) throw new Error("TOKVERA_API_BASE_URL is required");
  if (!sessionToken) throw new Error("TOKVERA_SESSION_TOKEN is required");
  if (!tenantId) throw new Error("TOKVERA_TENANT_ID is required");

  const windowRange = buildWindow(days);
  const params = new URLSearchParams({
    from: windowRange.from,
    to: windowRange.to,
  });
  if (projectId) params.set("project_id", projectId);

  const list = await requestJson({
    baseUrl,
    sessionToken,
    tenantId,
    path: `/v1/usage/customers?${params.toString()}`,
  });

  const selectedCustomer =
    list.items.find((item) => item.customer_id === preferredCustomerId) || list.items[0] || null;

  if (!selectedCustomer) {
    console.log("No customer-attributed usage found in this window.");
    return;
  }

  const detail = await requestJson({
    baseUrl,
    sessionToken,
    tenantId,
    path: `/v1/usage/customers/${encodeURIComponent(selectedCustomer.customer_id)}?${params.toString()}`,
  });

  const webhookPayload = {
    schema_version: "customer_usage.v1",
    event_type: "customer.usage.updated",
    customer_id: selectedCustomer.customer_id,
    credit_bucket: detail.by_credit_bucket?.[0]?.group || null,
    remaining_credits: null,
    threshold: null,
    metadata: detail.metadata || null,
    metadata_observed_at: detail.metadata_observed_at || null,
    overview: detail.overview,
  };

  console.log("customer usage sync example complete");
  console.log({
    window: windowRange,
    selected_customer: selectedCustomer.customer_id,
    ledger_snapshot: toLedgerSnapshot(selectedCustomer),
    webhook_projection: toWebhookLedgerEvent(webhookPayload),
    next_step:
      "Persist ledger_snapshot in your own billing store and validate webhook_projection against your webhook handler before enabling automated customer_usage webhooks.",
  });
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
